data_model/primitives/src/conststr.rs

```diff
@@ -0,0 +1,545 @@
+where
+    ConstStringData: AsRef<T>,
+{
+    #[inline]
+    fn as_ref(&self) -> &T {
+        self.0.as_ref()
+    }
+}
+
+impl<T: Into<ConstStringData>> From<T> for ConstString {
+    #[inline]
+    fn from(value: T) -> Self {
+        Self(value.into())
+    }
+}
+
+impl Display for ConstString {
+    fn fmt(&self, f: &mut Formatter<'_>) -> FmtResult {
+        write!(f, "{}", &**self)
+    }
+}
+
+impl Debug for ConstString {
+    fn fmt(&self, f: &mut Formatter<'_>) -> FmtResult {
+        write!(f, "{:?}", &**self)
+    }
+}
+
+impl Hash for ConstString {
+    #[inline]
+    fn hash<H: Hasher>(&self, state: &mut H) {
+        (**self).hash(state)
+    }
+}
+
+impl PartialOrd for ConstString {
+    #[inline]
+    fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
+        PartialOrd::partial_cmp(&**self, &**other)
+    }
+}
+
+impl Ord for ConstString {
+    #[inline]
+    fn cmp(&self, other: &Self) -> Ordering {
+        Ord::cmp(&**self, &**other)
+    }
+}
+
+impl PartialEq for ConstString {
+    #[inline]
+    fn eq(&self, other: &Self) -> bool {
+        PartialEq::eq(&**self, &**other)
+    }
+}
+
+macro_rules! impl_eq {
+    ($($ty:ty,)*) => {
+        impl_eq!($($ty),*);
+    };
+    ($($ty:ty),*) => {$(
+        impl PartialEq<$ty> for ConstString {
+            #[allow(clippy::string_slice, clippy::deref_by_slicing)]
+            #[inline]
+            fn eq(&self, other: &$ty) -> bool {
+                PartialEq::eq(&self[..], &other[..])
+            }
+        }
+
+        impl PartialEq<ConstString> for $ty {
+            #[allow(clippy::string_slice, clippy::deref_by_slicing)]
+            #[inline]
+            fn eq(&self, other: &ConstString) -> bool {
+                PartialEq::eq(&self[..], &other[..])
+            }
+        }
+    )*};
+}
+
+impl_eq!(String, str, &str);
+
+impl Eq for ConstString {}
+
+impl Serialize for ConstString {
+    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
+    where
+        S: Serializer,
+    {
+        serializer.serialize_str(self)
+    }
+}
+
+impl Encode for ConstString {
+    fn size_hint(&self) -> usize {
+        self.as_bytes().size_hint()
+    }
+
+    fn encode_to<W: Output + ?Sized>(&self, dest: &mut W) {
+        self.as_bytes().encode_to(dest)
+    }
+
+    fn encode(&self) -> Vec<u8> {
+        self.as_bytes().encode()
+    }
+
+    fn using_encoded<R, F: FnOnce(&[u8]) -> R>(&self, f: F) -> R {
+        self.as_bytes().using_encoded(f)
+    }
+}
+
+impl IntoSchema for ConstString {
+    fn type_name() -> String {
+        String::type_name()
+    }
+    fn schema(map: &mut MetaMap) {
+        String::schema(map);
+    }
+}
+
+/// Union representing const-string variants: inlined or boxed.
+/// Distinction between variants are achieved by setting least significant bit for inlined variant.
+/// Boxed variant should have have 4/3/2 (depending on architecture 64/32/16 bit pointer width) trailing zeros due to pointer alignment.
+#[repr(C)]
```

Comment:
why does it have to be `repr(C)`?

Reply:
Checking the LSB, so we can use a bit for  a tag, rather than a whole byte?

Reply:
from the looks of it, I don't think you need it for `ConstStringData` or `InlinedString` in the same way as you don't have it for `Box<str>` which I don't think has a defined repr

Reply:
ok, InlinedString has to be repr(C), but I don't see why this union would have to be as well. The compiler will add appropriate offsets when casting union into InlinedString and you will be accessing LSB of that offset

I would note that you're accessing bit's directly and those aren't a concept in rust. There are no bitfields as in C. There are some bitfield crates though which would make these bit manipulation more obvious and readable. We might consider using them

Reply:
ok, I get it. You are implicitly using tagged pointer when union variant is BoxedString. However, isn't it the case that the MSB, not LSB, is unset in every pointer? If tagged pointers is indeed used, another issue is that I think that the representation of boxed slice Box<str> is undefined because slices don't have a defined representation. If this is the case you should use raw pointers

I would note that you're accessing bit's directly and those aren't a concept in rust. There are no bitfields as in C. There are some bitfield crates though which would make these bit manipulation more obvious and readable. We might consider using them

Reply:
Strictly speaking there aren't bitfields on any CPU architecture regardless of programming language. The smallest addressable unit is always the byte, and bit-packing operations almost always get implemented with masks and binary logic operators. 

With that said, using the bitfield crates could make your  code more readable, as would better documentation (preferably both).

Reply:
> You are implicitly using tagged pointer when union variant is BoxedString

Yeah, `BoxedString` is tagged with zero at the  LSB.

> However, isn't it the case that the MSB, not LSB, is unset in every pointer?

Both options are possible:
- in case of MSB it comes from fact that on x86 we have 48bit logical addressing space;
- in case of LSB it happens because of alignment. 

Useful link: [pointer packing](https://www.1024cores.net/home/lock-free-algorithms/tricks/pointer-packing).

> another issue is that I think that the representation of boxed slice Box<str> is undefined because slices don't have a defined representation. If this is the case you should use raw pointers

You are right, my bad, i assumed layout of `Box<str>` would be pointer and length when debugging and run simple **[test](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=a0cb46112d554c70331d6dfd2c4cd802),** i'm going to switch to pointers in `BoxedString` struct, `Box::into_raw` should work (strore `NotNull<u8>` and length).

@appetrosyan, @mversic, I made wrong assumption, i think that if `*mut u8` has minimal alignment of 8 bytes, it means that pointer would have 3 last bits equal to zero, but in fact it means that reference to `*mut u8` would have 3 last bits equal to zero.

Tests passes because `libc::malloc` had minimal alignment of 16 bytes on `x86_64` ([MALLOC_ALIGNMENT](https://sourceware.org/glibc/wiki/MallocInternals).)

Windows `HeapAlloc` have the same alignment ([MEMORY_ALLOCATION_ALIGNMENT](https://docs.microsoft.com/en-us/windows/win32/api/heapapi/nf-heapapi-heapalloc)).

However i'm not sure about `wee_alloc`.

I have few options:
- Leave as is and document.
- Rely on MSB to be equal to 0;
- Alloc memory for `BoxedString` with at least `u16` alignment, so it would have LSB guaranteed to be 0 (most reliable).

Reply:
Could you try it with jemalloc? We might need to statically link an allocator, becauseof musl, so this might be a non-issue.

Reply:
I like your idea of storing the tag in the MSB of len, because with this approach we store the tag in the same way for the Boxed and Inlined variants.

Reply:
> Could you try it with jemalloc? We might need to statically link an allocator, becauseof musl, so this might be a non-issue.

It works with jemmalloc just fine.

Reply:
Good. We can add support for other allocators later if we need to.